pallets/utxo/tokens/src/lib.rs

```diff
@@ -5,24 +5,33 @@ use serde::{Deserialize, Serialize};
 
 use codec::{Decode, Encode};
 use frame_support::{dispatch::Vec, RuntimeDebug};
+use sp_core::{sr25519::Public, H256};
 
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, Hash)]
-pub struct TokenInstance {
-    pub id: u64,
-    pub name: Vec<u8>,
-    pub ticker: Vec<u8>,
-    pub supply: u128,
-    // We can add another fields like:
-    //      pub number_format: NumberFormat,
-    //      pub image: UUID,
-    //      pub transaction: XXX,
+pub enum TokenInstance {
+    Normal {
+        id: H256,
+        name: Vec<u8>,
```

Comment:
I wonder if it's worth stripping the name out of the tx to reduce the size and just let the UI side handle this?

Reply:
What do you think @TheQuantumPhysicist ? I can remove all tokens names, tickers, URLs from transactions but how we shall mapping the blockchain data with these?

Reply:
So you don't necessarily need to remove it all, but ticker and name seems redundant

Reply:
I think all this information (name, ticker, etc) should be part of a unified interface that we extract from data. The only thing that should be identifying a token, in general, is its `id`. Everything else should be defined by the user and/or the interface they choose to implement in their NFT.

Reply:
> I think all this information (name, ticker, etc) should be part of a unified interface that we extract from data. The only thing that should be identifying a token, in general, is its `id`. Everything else should be defined by the user and/or the interface they choose to implement in their NFT.

May you please give me some examples? For me it is not clear.

Reply:
What means `defined by the user and/or the interface`?

It's like a third-party website where artists could paint their pictures and there stored all information about NFT but in our blockchain, we store just ID and hash of the picture, is it?

Reply:
Looking at Ethereum, the components of a transaction are destination, nonce, gasLimit, gasPrice and data. That's all. We don't need nonce, because we have UTXOs. The gas thing doesn't exist on substrate because they use weights instead (we'll have a discussion about that). And then, everyyyyyyything else should go into data. Notice that "data" can go into smart contracts, but in our case, we want to create a semi smart contract system where the data of a transaction is interpreted from the data in it instead of executing a smart contract. You start with an NFT trait/interface (which has at least the data and version method in it), and then you create NFTv1 trait that implements it and adds the fields you wanna add. All should be serializable within data. This should also contain the serialization version. So that when you create NFTv2 (which also implements NFT trait), it should be ready to be read. The serialization/deserialization should be customized.

A more general format of interfaces would be something like Ethereum's, where an interface is created as a json object that specifies what functions should be extracted from the data and its signature, but this can only be made after we define a format for the data in general.

To be clear, I'm not saying here how you should implement it. But what I'm saying is that you should not have any fixed fields whatsoever within your most basic NFT struct. Only `data` and `id` should be present. Everything else should be extracted dynamically based on the version.

Reply:
@TheQuantumPhysicist I think, Ben's question was - The `TokenInstance::Normal` doesn't relate to NFT it's standard tokens without `data` field. But for this kind of token, we save their names and tickers. Do we need to do that?

Reply:
Nope. Same thing. Whether they're NFT or not, everything should be inside the data except for the id. The serialization format of the data should be used to extract token information.

Reply:
> Nope. Same thing. Whether they're NFT or not, everything should be inside the data except for the id. The serialization format of the data should be used to extract token information.

Good, why we can't use just Enum instead of the version and data fields? Anyway, we are deriving Encode and Decode from the SCALE